reference-implementation/__tests__/parsing-addresses.js

```diff
@@ -49,14 +49,14 @@ describe('Relative URL-like addresses', () => {
       }`,
       'https://base.example/path1/path2/path3',
       {
-        dotSlash: [expect.toMatchURL('https://base.example/path1/path2/')],
-        dotDotSlash: [expect.toMatchURL('https://base.example/path1/')],
-        slash: [expect.toMatchURL('https://base.example/')]
+        dotSlash: ['https://base.example/path1/path2/'],
+        dotDotSlash: ['https://base.example/path1/'],
+        slash: ['https://base.example/']
       }
     );
   });
 
-  it('should ignore percent-encoded variants of ./, ../, or /', () => {
+  it('should treat percent-encoded variants of ./, ../, or / as non-URLs', () => {
```

Comment:
So now we have bare specifiers in the output. This is not great for the ultimate, composed import map; ideally we'd give people a warning in such cases. I.e. it'd be ideal if bare specifiers in the first import map were discarded with a warning, bare specifiers in the second that did not map to anything from the first were discarded with a warning, etc.

I'm not sure whether that's feasible with the currently-envisioned architecture. (This is part of why I anticipated this being an even bigger change than it currently is.) I don't necessarily think we should demand it, since it's just about good warnings at parse time (instead of errors at use time), but we should weigh it.

Of course, if there's some nice way to make that work with minimal disruption, that'd be great as well.

Reply:
We should be able to give warnings for this and remove the specifiers from the output, but it would happen at composition time, not parse time. (You can conceptualize this as starting the document with an empty map and composing any that are present, in order.)

Reply:
To spell this out a little more:

You can't give a warning for a bare specifier on the right-hand side _at parse time_, because that can be legal when adding an import map to a page which provides a mapping for that specifier. For example,

```html
<script type="importmap">
{
  "imports": {
    "moment": "/node_modules/moment/src/moment.js"
  }
}
</script>
<script type="importmap">
{
  "imports": {
    "m": "moment"
  }
}
</script>
```
is fine, despite the second map having the bare specifier `"moment"` on the RHS, because that `"moment"` is mapped to a real URL by the first map.

But you _can_ give a warning for a bare specifier on the RHS which is still present _after composition_, since maps only cascade to earlier maps, not later ones. And you can use this mechanism to give warnings for the first map on the page by acting as if every page starts out having the empty import map `{}` loaded already. For example,

```html
<script type="importmap">
{
  "imports": {
    "moment": "/node_modules/moment/src/moment.js",
    "lodash": "lodash/index.js"
  }
}
</script>
<script type="importmap">
{
  "imports": {
    "jquery": "javascript:bad",
    "l": "lodash"
  }
}
</script>
```

can give a warning for `"lodash"` when adding the first map and for both `"jquery"` and `"l"` when adding the second map, since you know when each map is added to the page that those point to bare specifiers. The PR has been updated with this behavior.